File: moxy-compiler/src/main/java/com/omegar/mvp/compiler/pipeline/UniqueValidator.java
package com.omegar.mvp.compiler.pipeline;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Anton Knyazev on 04.12.2020.
 */
public class UniqueValidator<T> extends Validator<T> {

    private final List<T> mList = new ArrayList<>();

    @Override
    public synchronized boolean validate(T input) {
        if (mList.contains(input)) {
            return false;
        }
        mList.add(input);
        return true;
    }

    @Override
    protected void finish(PipelineContext<T> nextContext) {
        super.finish(nextContext);
        mList.clear();
    }

}

File: moxy/compiler/src/main/java/com/omegar/mvp/compiler/pipeline/UniqueValidator.kt
package com.omegar.mvp.compiler.pipeline

/**
 * Created by Anton Knyazev on 04.12.2020.
 */
class UniqueValidator<T> : Validator<T>() {

    private val list: MutableList<T> = ArrayList()

    override fun validate(input: T): Boolean {
        if (list.contains(input)) {
            return false
        }
        list.add(input)
        return true
    }

    override fun finish(nextContext: PipelineContext<T>?) {
        super.finish(nextContext)
        list.clear()
    }
}